types/src/block/block.rs

```diff
@@ -57,22 +56,25 @@ impl<P: NonceProvider> BlockBuilder<P> {
         self
     }
 
-    /// Adds a nonce provider to a [`BlockBuilder`].
+    /// Adds a nonce to a [`BlockBuilder`].
     #[inline(always)]
-    pub fn with_nonce_provider(mut self, nonce_provider: P) -> Self {
-        self.nonce_provider = Some(nonce_provider);
+    pub fn with_nonce(mut self, nonce: u64) -> Self {
+        self.nonce = Some(nonce);
         self
     }
 
     /// Finishes the [`BlockBuilder`] into a [`Block`].
-    pub fn finish(self, min_pow_score: u32) -> Result<Block, Error> {
+    pub fn finish_nonce_provider<F: Fn(&[u8]) -> Result<u64, PoWError>>(
```

Comment:
I actually don't really like having 2 finish, we should be able to error if no nonce or no nonce provider has been given, instead of putting a default 0 nonce which doesn't really make sense

Reply:
Well, the only solution would have been to box the provider but we can't have a Clone requirement on boxed traits, so the builder couldn't be Clone anymore. Forget about it.